File: src/hotkeys.ts
import { parseKeys, matchKey, type KeyEvent } from "./key-parser.js";
import { debug } from "./debug.js";

const TIMEOUT_MS = 1000;

export type HotkeyAction =
  | { type: "dashboard" }
  | { type: "focus"; index: number }
  | { type: "next" }
  | { type: "prev" }
  | { type: "create" }
  | { type: "kill" }
  | { type: "passthrough"; data: string };

export type ActionCallback = (action: HotkeyAction) => void;

export class HotkeyHandler {
  private waitingForAction = false;
  private timeout: ReturnType<typeof setTimeout> | null = null;
  private onAction: ActionCallback;

  constructor(onAction: ActionCallback) {
    this.onAction = onAction;
  }

  /**
   * Process raw stdin data. Returns any bytes that should be
   * forwarded to the active PTY (non-hotkey data).
   */
  feed(data: Buffer): string | null {
    const events = parseKeys(data);
    debug(`hotkey feed: ${events.length} events, waiting=${this.waitingForAction}, bytes=${data.length}`, "hotkey");
    if (events.length === 0) return data.toString();

    // If we're waiting for the action key after leader
    if (this.waitingForAction) {
      this.clearTimeout();
      this.waitingForAction = false;
      this.hideLeaderIndicator();
      return this.handleActionEvent(events[0], data);
    }

    // Check if this is the leader key (Ctrl+A)
    if (events.length === 1 && matchKey(events[0], "ctrl+a")) {
      this.waitingForAction = true;
      this.showLeaderIndicator();
      this.timeout = setTimeout(() => {
        this.waitingForAction = false;
        this.hideLeaderIndicator();
        // Timeout: forward the original data
        this.onAction({ type: "passthrough", data: data.toString() });
      }, TIMEOUT_MS);
      return null;
    }

    // Regular data — pass through
    return data.toString();
  }

  private handleActionEvent(event: KeyEvent, rawData: Buffer): string | null {
    // Double Ctrl+A: send literal through
    if (matchKey(event, "ctrl+a")) {
      this.onAction({ type: "passthrough", data: rawData.toString() });
      return null;
    }

    const key = event.name || event.char;

    switch (key) {
      case "d":
        this.onAction({ type: "dashboard" });
        return null;
      case "n":
        this.onAction({ type: "next" });
        return null;
      case "p":
        this.onAction({ type: "prev" });
        return null;
      case "c":
        this.onAction({ type: "create" });
        return null;
      case "x":
        this.onAction({ type: "kill" });
        return null;
      default:
        // Check for digits 1-9
        if (key >= "1" && key <= "9") {
          this.onAction({ type: "focus", index: parseInt(key) - 1 });
          return null;
        }
        // Unknown action key — forward raw data
        this.onAction({ type: "passthrough", data: rawData.toString() });
        return null;
    }
  }

  private showLeaderIndicator(): void {
    // Save cursor, move to top-right, show indicator, restore cursor
    const cols = process.stdout.columns ?? 80;
    const label = " ^A → ? ";
    const col = cols - label.length;
    process.stdout.write(
      `\x1b7` +                              // save cursor
      `\x1b[1;${col}H` +                     // move to top-right
      `\x1b[7;33m${label}\x1b[0m` +          // inverse yellow
      `\x1b8`                                 // restore cursor
    );
  }

  private hideLeaderIndicator(): void {
    // Save cursor, clear the indicator area, restore cursor
    const cols = process.stdout.columns ?? 80;
    const label = " ^A → ? ";
    const col = cols - label.length;
    process.stdout.write(
      `\x1b7` +
      `\x1b[1;${col}H` +
      `${" ".repeat(label.length)}` +
      `\x1b8`
    );
  }

  private clearTimeout(): void {
    if (this.timeout) {
      clearTimeout(this.timeout);
      this.timeout = null;
    }
  }

  destroy(): void {
    this.clearTimeout();
  }
}

File: native/crates/aimux/src/hotkeys.rs
use crate::terminal_key_parser::{KeyEvent, command_key, match_key, parse_keys};
use serde_json::{Value, json};

const HOTKEY_TIMEOUT_MS: i64 = 1_000;

pub fn run_hotkeys_contract_case(input: &Value) -> Value {
    let columns = input["columns"].as_u64().unwrap_or(80) as usize;
    let mut handler = HotkeyHandler::new(columns);
    let returns = input["chunks"]
        .as_array()
        .into_iter()
        .flatten()
        .map(|chunk| {
            let raw = chunk.as_str().unwrap_or_default();
            handler
                .feed(raw.as_bytes())
                .map_or(Value::Null, Value::String)
        })
        .collect::<Vec<_>>();
    let _ = HOTKEY_TIMEOUT_MS;
    json!({
        "returns": returns,
        "actions": handler.actions,
        "writes": handler.writes,
    })
}

#[derive(Debug, Clone)]
pub struct HotkeyHandler {
    waiting_for_action: bool,
    columns: usize,
    actions: Vec<Value>,
    writes: Vec<String>,
}

impl HotkeyHandler {
    pub fn new(columns: usize) -> Self {
        Self {
            waiting_for_action: false,
            columns,
            actions: Vec::new(),
            writes: Vec::new(),
        }
    }

    pub fn feed(&mut self, data: &[u8]) -> Option<String> {
        let events = parse_keys(data);
        if events.is_empty() {
            return Some(String::from_utf8_lossy(data).into_owned());
        }
        if self.waiting_for_action {
            self.waiting_for_action = false;
            self.hide_leader_indicator();
            return self.handle_action_event(&events[0], data);
        }
        if events.len() == 1 && match_key(&events[0], "ctrl+a") {
            self.waiting_for_action = true;
            self.show_leader_indicator();
            return None;
        }
        Some(String::from_utf8_lossy(data).into_owned())
    }

    fn handle_action_event(&mut self, event: &KeyEvent, raw_data: &[u8]) -> Option<String> {
        if match_key(event, "ctrl+a") {
            self.actions.push(json!({
                "type": "passthrough",
                "data": String::from_utf8_lossy(raw_data).into_owned(),
            }));
            return None;
        }
        let key = if event.name.is_empty() {
            event.char.as_str()
        } else {
            event.name.as_str()
        };
        let lower_key = command_key(event);
        if key == "W" {
            self.actions.push(json!({ "type": "worktree-list" }));
            return None;
        }
        if key == "P" {
            self.actions.push(json!({ "type": "work-outline" }));
            return None;
        }
        match lower_key.as_str() {
            "d" => self.actions.push(json!({ "type": "dashboard" })),
            "?" => self.actions.push(json!({ "type": "help" })),
            "n" => self.actions.push(json!({ "type": "next" })),
            "i" => self.actions.push(json!({ "type": "coordination" })),
            "p" => self.actions.push(json!({ "type": "prev" })),
            "c" => self.actions.push(json!({ "type": "create" })),
            "o" => self.actions.push(json!({ "type": "create-overseer" })),
            "x" => self.actions.push(json!({ "type": "kill" })),
            "s" => self.actions.push(json!({ "type": "switcher" })),
            "w" => self.actions.push(json!({ "type": "worktree-create" })),
            "v" => self.actions.push(json!({ "type": "review" })),
            _ if ("1"..="9").contains(&key) => {
                let index = key.parse::<i64>().unwrap_or(1) - 1;
                self.actions
                    .push(json!({ "type": "focus", "index": index }));
            }
            _ => self.actions.push(json!({
                "type": "passthrough",
                "data": String::from_utf8_lossy(raw_data).into_owned(),
            })),
        }
        None
    }

    fn show_leader_indicator(&mut self) {
        let label = " ^A \u{2192} ? ";
        let col = self.columns.saturating_sub(label.chars().count());
        self.writes
            .push(format!("\x1b7\x1b[1;{col}H\x1b[7;33m{label}\x1b[0m\x1b8"));
    }

    fn hide_leader_indicator(&mut self) {
        let label = " ^A \u{2192} ? ";
        let col = self.columns.saturating_sub(label.chars().count());
        self.writes.push(format!(
            "\x1b7\x1b[1;{col}H{}\x1b8",
            " ".repeat(label.chars().count())
        ));
    }
}
